src/main/scala/io/iohk/ethereum/consensus/pow/validators/OmmersValidator.scala

```diff
@@ -30,11 +30,16 @@ trait OmmersValidator {
   )(implicit blockchainConfig: BlockchainConfig): Either[OmmersError, OmmersValid] = {
 
     val getBlockHeaderByHash: ByteString => Option[BlockHeader] = blockchainReader.getBlockHeaderByHash
-    val bestBranch = blockchainReader.getBestBranch()
     val getNBlocksBack: (ByteString, Int) => List[Block] =
       (_, n) =>
-        ((blockNumber - n) until blockNumber).toList
-          .flatMap(nb => bestBranch.getBlockByNumber(nb))
+        Iterator
+          .iterate(blockchainReader.getBlockByHash(parentHash))(
+            _.flatMap(block => blockchainReader.getBlockByHash(block.header.parentHash))
+          )
+          .take(n)
+          .collect { case Some(block) => block }
+          .toList
```

Comment:
The refactor doesn't change this point, so consider it an open question: do we already have an idea of the number of lists and their size, is there a memory risk somehow ?

Reply:
Yes, the list is actually bounded to 6 elements. So there should not be a memory problem.
We want to check that a given block was not already used as an ommer in the parents blocks, and an ommer is only valid for 6 generations, so don't have to go back further than that.